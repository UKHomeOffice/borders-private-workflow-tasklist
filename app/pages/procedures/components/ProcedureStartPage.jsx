import * as React from "react";
import PropTypes from 'prop-types';
import StartForm from "../../../core/start-forms/components/StartForm";
import {isFetchingProcessDefinition, processDefinition} from "../selectors";
import {bindActionCreators} from "redux";
import {createStructuredSelector} from "reselect";
import * as actions from "../actions";
import ImmutablePropTypes from "react-immutable-proptypes";
import {connect} from "react-redux";
import {withRouter} from "react-router";
import queryString from 'query-string';
import {submissionToWorkflowSuccessful, submittingToWorkflow} from "../../../core/start-forms/selectors";
import Spinner from 'react-spinkit';
import { DataSpinner } from '../../../core/components/DataSpinner';

class ProcessStartPage extends React.Component {

    componentDidMount() {
        const params = queryString.parse(this.props.location.search);
        this.props.fetchProcessDefinition(params.processKey);
    }

    componentWillUnmount() {
        this.props.reset();
    }


    render() {
        const {isFetchingProcessDefinition, processDefinition, submittingToWorkflow} = this.props;
        const pointerStyle = {cursor: 'pointer', paddingTop: '10px', textDecoration: 'underline'};
        return <div>
            <div style={pointerStyle} onClick={(event) => this.props.history.replace('/procedures')}>Back to
                procedures
            </div>
            <div className="grid-row">
                {submittingToWorkflow ?
                    <div style={{display: 'flex', justifyContent: 'center', paddingTop: '20px'}}><DataSpinner
                       message="Starting procedure..."/></div> : <div/>
                }
                <div className="column-full">
                    <fieldset>
                        {isFetchingProcessDefinition ? <div>Loading form...</div> : <div>
                            {processDefinition ? <div>

                                <h2 className="heading-large">
                    <span
                        className="heading-secondary">Operational procedure</span> {processDefinition.getIn(['process-definition', 'name'])}
                                </h2>


                                <StartForm formName={processDefinition.get('formKey')}
                                           processKey={processDefinition.getIn(['process-definition', 'key'])}
                                           processName={processDefinition.getIn(['process-definition', 'name'])}
                                           {...this.props}/>
                            </div> : <div>
                                No process definition found
                            </div>}

                        </div>
                        }

                    </fieldset>
                </div>
            </div>
        </div>
    };

}

ProcessStartPage.propTypes = {
    fetchProcessDefinition: PropTypes.func.isRequired,
    processDefinition: ImmutablePropTypes.map,
    isFetchingProcessDefinition: PropTypes.bool,
    submittingToWorkflow: PropTypes.bool,
    submissionToWorkflowSuccessful: PropTypes.bool
};

const mapStateToProps = createStructuredSelector({
    processDefinition: processDefinition,
    isFetchingProcessDefinition: isFetchingProcessDefinition,
    submittingToWorkflow: submittingToWorkflow,
    submissionToWorkflowSuccessful: submissionToWorkflowSuccessful,

});

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(ProcessStartPage));
