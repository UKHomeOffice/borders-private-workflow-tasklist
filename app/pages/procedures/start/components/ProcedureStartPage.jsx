import * as React from 'react';
import PropTypes from 'prop-types';
import {form, isFetchingProcessDefinition, loadingForm, processDefinition, submissionStatus,} from '../selectors';
import {bindActionCreators} from 'redux';
import * as actions from '../actions';
import ImmutablePropTypes from 'react-immutable-proptypes';
import {connect} from 'react-redux';
import {withRouter} from 'react-router';
import AppConstants from '../../../../common/AppConstants';

import DataSpinner from '../../../../core/components/DataSpinner';
import Loader from 'react-loader-advanced';
import StartForm from './StartForm';
import NotFound from '../../../../core/components/NotFound';
import secureLocalStorage from '../../../../common/security/SecureLocalStorage';
import withLog from '../../../../core/error/component/withLog';
import {FAILED, SUBMISSION_SUCCESSFUL, SUBMITTING} from '../constants';


export class ProcessStartPage extends React.Component {

    constructor(props) {
        super(props);
        this.secureLocalStorage = secureLocalStorage;
        this.handleSubmission = this.handleSubmission.bind(this);
    }

    componentDidMount() {
        if (this.props.processKey) {
            this.props.fetchProcessDefinition(this.props.processKey);
        } else {
            const {match: {params}} = this.props;
            this.props.fetchProcessDefinition(params.processKey);
        }
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        if (this.props.isFetchingProcessDefinition !== prevProps.isFetchingProcessDefinition
            && !this.props.isFetchingProcessDefinition) {
            this.props.fetchForm(this.props.processDefinition.get('formKey'));
        }

        const {submissionStatus} = this.props;
        if (submissionStatus !== prevProps.submissionStatus && (this.form && this.form.formio)) {
            const path = this.props.history.location.pathname;
            const user = this.props.kc.tokenParsed.email;
            const processKey = this.props.processDefinition.getIn(['process-definition', 'key']);
            this.handleSubmission(submissionStatus, user, path, processKey);
        }
    }

    handleSubmission(submissionStatus, user, path, processKey) {
        switch (submissionStatus) {
            case SUBMITTING :
                this.props.log([{
                    user: user,
                    path: path,
                    level: 'info',
                    message: `Submitting data for ${processKey}`
                }]);
                break;
            case SUBMISSION_SUCCESSFUL :
                window.scrollTo(0, 0);
                this.form.formio.emit('submitDone');
                this.secureLocalStorage.removeAll();
                this.props.log([{
                    user: user,
                    path: path,
                    level: 'info',
                    message: `Procedure ${processKey} successfully started`
                }]);
                if (this.props.redirectPath) {
                    this.props.history.replace(this.props.redirectPath);
                } else {
                    this.props.history.replace(AppConstants.YOUR_TASKS_PATH);
                }
                break;
            case FAILED :
                this.props.log([{
                    user: user,
                    path: path,
                    level: 'error',
                    message: `Procedure ${processKey} failed to be initiated`
                }]);
                const submission = this.form.formio.submission ? this.form.formio.submission :
                    this.secureLocalStorage.get(this.props.processDefinition.getIn(['process-definition', 'id']));
                this.form.formio.emit('error');
                this.form.formio.emit('change', submission);
                break;
            default:
                this.props.log([{
                    level: 'warn',
                    path: path,
                    user: user,
                    message: 'Unknown submission status',
                    submissionStatus: submissionStatus
                }]);
        }
    }

    componentWillUnmount() {
        this.props.clearProcessDefinition();
    }

    handleCustomEvent = (event) => {
        switch (event.type) {
            case 'cancel':
                this.secureLocalStorage.removeAll();
                this.props.history.replace(AppConstants.DASHBOARD_PATH);
                break;
            case 'save-draft':
                break;
            default:
        }
    };

    render() {
        const {processDefinition, submissionStatus, form, loadingForm} = this.props;
        if (loadingForm) {
            return <DataSpinner message="Loading procedure..."/>;
        } else {
            if (!form) {
                return <NotFound resource="Form" id={processDefinition.get('formKey')}/>;
            }
            const procedureKey = processDefinition.getIn(['process-definition', 'key']);
            const variableInput = form.components.find(c => c.key === 'submitVariableName');
            const variableName = variableInput ? variableInput.defaultValue : form.name;
            const process = processDefinition.getIn(['process-definition', 'name']) ?
                processDefinition.getIn(['process-definition', 'name']) : procedureKey;

            return <div>
                {!this.props.noBackLink ? <a href="#" className="govuk-back-link govuk-!-font-size-19"
                                               style={{textDecoration: 'none'}}
                                               onClick={(event) => {
                                                   event.preventDefault();
                                                   this.props.history.replace('/procedures')
                                               }}>Back to procedures</a> : null}

                <Loader show={submissionStatus === SUBMITTING}
                        message={<div style={{
                            justifyContent: 'center',
                            zIndex: 100,
                            borderRadius: 0,
                            position: 'absolute',
                            left: 0,
                            right: 0,
                            top: '-20px',
                            margin: 'auto'
                        }}><DataSpinner
                            message="Starting procedure..."/></div>}
                        hideContentOnLoad={submissionStatus === SUBMITTING}
                        foregroundStyle={{color: 'black'}}
                        backgroundStyle={{backgroundColor: 'white'}}>
                    <div className="govuk-grid-row">
                        <div className="govuk-grid-column-full">
                            <div>
                                <span className="govuk-caption-l">Operational procedure</span>
                                <h2 className="govuk-heading-l">{processDefinition.getIn(['process-definition', 'name'])}</h2>

                                <StartForm {...this.props}
                                           startForm={form}
                                           formReference={(formLoaded) => {
                                               this.form = formLoaded;
                                               if (this.form) {
                                                   this.form.createPromise.then(() => {
                                                       this.form.formio.on('componentError', (error) => {
                                                           const path = this.props.history.location.pathname;
                                                           const user = this.props.kc.tokenParsed.email;
                                                           this.props.log([{
                                                               user: user,
                                                               path: path,
                                                               level: 'error',
                                                               form: {
                                                                   name: form.name,
                                                                   path: form.path,
                                                                   display: form.display
                                                               },
                                                               message: error.message,
                                                               component: {
                                                                   label: error.component.label,
                                                                   key: error.component.key
                                                               }
                                                           }]);
                                                       })
                                                   })
                                               }

                                           }}
                                           removeAll={this.secureLocalStorage.removeAll()}
                                           submission={this.secureLocalStorage.get(processDefinition.getIn(['process-definition', 'id']))}
                                           dataChange={(instance) => {
                                               this.secureLocalStorage.set(processDefinition.getIn(['process-definition', 'id']), instance.data)
                                           }
                                           }
                                           onCustomEvent={(event) => this.handleCustomEvent(event)}
                                           handleSubmit={(submission) => {
                                               if (this.form && this.form.formio) {
                                                   this.form.formio.submitted = true;
                                                   this.form.formio.submitting = true;
                                               }
                                               this.props.submit(form._id, procedureKey,
                                                   variableName,
                                                   submission.data, process,
                                                   this.props.nonShiftApiCall);
                                           }}
                                />
                            </div>
                        </div>
                    </div>
                </Loader>
            </div>;
        }

    };

}

ProcessStartPage.propTypes = {
    log: PropTypes.func,
    fetchForm: PropTypes.func,
    fetchProcessDefinition: PropTypes.func.isRequired,
    clearProcessDefinition: PropTypes.func.isRequired,
    submit: PropTypes.func,
    processDefinition: ImmutablePropTypes.map,
    submissionStatus: PropTypes.string,
    noBackLink: PropTypes.bool
};

const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export default withRouter(connect((state) => {
    return {
        processDefinition: processDefinition(state),
        submissionStatus: submissionStatus(state),
        form: form(state),
        loadingForm: loadingForm(state),
        isFetchingProcessDefinition: isFetchingProcessDefinition(state),
        kc: state.keycloak
    };
}, mapDispatchToProps)(withLog(ProcessStartPage)));
