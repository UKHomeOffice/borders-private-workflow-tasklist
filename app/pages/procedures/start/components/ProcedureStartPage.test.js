import { mount } from 'enzyme/build';
import {ProcessStartPage} from './ProcedureStartPage';
import React from 'react';
import Immutable from 'immutable';

describe('Start a procedure page', () => {

  it('renders loading bar if form is loading', async() => {
    const props = {
      loadingForm: true,
      submittingToWorkflow: false,
      submissionToWorkflowSuccessful: false,
      isFetchingProcessDefinition: false,
      match : {
        params: {
          "processKey": "processKey"
        }
      }
    };
    const fetchProcessDefinition = jest.fn();
    const clearProcessDefinition = jest.fn();

    const wrapper = await mount(<ProcessStartPage
      {...props}
      clearProcessDefinition={clearProcessDefinition}
      fetchProcessDefinition={fetchProcessDefinition}
    />);

    expect(fetchProcessDefinition).toHaveBeenCalled();
    expect(wrapper.find('#dataSpinner').exists()).toEqual(true);
    expect(wrapper.find('.loader-message').text()).toEqual('Loading procedure...');

  });

  it ('displays resource not found if form is missing', async() => {
    const props = {
      loadingForm: false,
      submittingToWorkflow: false,
      submissionToWorkflowSuccessful: false,
      isFetchingProcessDefinition: false,
      form: null,
      processDefinition: Immutable.fromJS({
          "formKey" : "formKey"
      }),
      match : {
        params: {
          "processKey": "processKey"
        }
      }
    };
    const fetchProcessDefinition = jest.fn();
    const clearProcessDefinition = jest.fn();

    const wrapper = await mount(<ProcessStartPage
      {...props}
      clearProcessDefinition={clearProcessDefinition}
      fetchProcessDefinition={fetchProcessDefinition}
    />);
    expect(fetchProcessDefinition).toHaveBeenCalled();

    expect(wrapper.find('div').text()).toEqual('Form with identifier formKey was not found')
  });
  it ('fetches form after loading process definition', async() => {
    const props = {
      loadingForm: true,
      submittingToWorkflow: false,
      submissionToWorkflowSuccessful: false,
      isFetchingProcessDefinition: true,
      form: null,
      processDefinition: Immutable.fromJS({
        "formKey" : "formKey"
      }),
      match : {
        params: {
          "processKey": "processKey"
        }
      }
    };
    const fetchProcessDefinition = jest.fn();
    const clearProcessDefinition = jest.fn();
    const fetchForm = jest.fn();

    const wrapper = await mount(<ProcessStartPage
      {...props}
      clearProcessDefinition={clearProcessDefinition}
      fetchProcessDefinition={fetchProcessDefinition}
      fetchForm={fetchForm}
    />);
    expect(fetchProcessDefinition).toHaveBeenCalled();
    wrapper.setProps({
      isFetchingProcessDefinition: false
    });
    expect(fetchForm).toHaveBeenCalled();
  });

  it ('renders the form and process definition', async() => {
    const props = {
      loadingForm: false,
      submittingToWorkflow: false,
      submissionToWorkflowSuccessful: false,
      isFetchingProcessDefinition: false,
      form : {
        "display": "form",
        "components": [
          {
            "id": 'firstNameId',
            "label": "First name",
            "allowMultipleMasks": false,
            "showWordCount": false,
            "showCharCount": false,
            "tableView": true,
            "alwaysEnabled": false,
            "type": "textfield",
            "input": true,
            "key": "firstName",
            "widget": {
              "type": ""
            }
          },
          {
            "id" : "surnameId",
            "label": "Surname",
            "allowMultipleMasks": false,
            "showWordCount": false,
            "showCharCount": false,
            "tableView": true,
            "alwaysEnabled": false,
            "type": "textfield",
            "input": true,
            "key": "surname",
            "widget": {
              "type": ""
            }
          },
          {
            "id" : "submitId",
            "type": "button",
            "label": "Submit",
            "key": "submit",
            "disableOnInvalid": true,
            "theme": "primary",
            "input": true,
            "tableView": true
          }
        ],
      },
      processDefinition: Immutable.fromJS({
        "formKey" : "formKey",
        "process-definition" : {
          "name" : "procedure"
        }
      }),
      match : {
        params: {
          "processKey": "processKey"
        }
      }
    };
    const fetchProcessDefinition = jest.fn();
    const clearProcessDefinition = jest.fn();

    const wrapper = await mount(<ProcessStartPage
      {...props}
      clearProcessDefinition={clearProcessDefinition}
      fetchProcessDefinition={fetchProcessDefinition}
    />);
    console.log(wrapper.html());
    expect(fetchProcessDefinition).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders Loader if submitted form', async() => {
    const props = {
      loadingForm: false,
      submittingToWorkflow: false,
      submissionToWorkflowSuccessful: false,
      isFetchingProcessDefinition: false,
      form : {
        "display": "form",
        "components": [
          {
            "id": 'firstNameId',
            "label": "First name",
            "allowMultipleMasks": false,
            "showWordCount": false,
            "showCharCount": false,
            "tableView": true,
            "alwaysEnabled": false,
            "type": "textfield",
            "input": true,
            "key": "firstName",
            "widget": {
              "type": ""
            }
          },
          {
            "id" : "surnameId",
            "label": "Surname",
            "allowMultipleMasks": false,
            "showWordCount": false,
            "showCharCount": false,
            "tableView": true,
            "alwaysEnabled": false,
            "type": "textfield",
            "input": true,
            "key": "surname",
            "widget": {
              "type": ""
            }
          },
          {
            "id" : "submitId",
            "type": "button",
            "label": "Submit",
            "key": "submit",
            "disableOnInvalid": true,
            "theme": "primary",
            "input": true,
            "tableView": true
          }
        ],
      },
      processDefinition: Immutable.fromJS({
        "formKey" : "formKey"
      }),
      match : {
        params: {
          "processKey": "processKey"
        }
      }
    };
    const fetchProcessDefinition = jest.fn();
    const clearProcessDefinition = jest.fn();

    const wrapper = await mount(<ProcessStartPage
      {...props}
      clearProcessDefinition={clearProcessDefinition}
      fetchProcessDefinition={fetchProcessDefinition}
    />);

    wrapper.setProps({submittingToWorkflow: true});
    console.log(wrapper.html());
    const loaderContent = wrapper.find('.Loader__content');
    expect(loaderContent.exists()).toEqual(true);
    expect(loaderContent.prop('style')).toEqual({opacity: 0});
  });

});
