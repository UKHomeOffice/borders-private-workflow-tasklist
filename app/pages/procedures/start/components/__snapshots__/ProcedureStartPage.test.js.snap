// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submit a form page renders the form and process definition 1`] = `
<div>
  <a
    className="govuk-back-link govuk-!-font-size-19"
    href="#"
    onClick={[Function]}
    style={
      Object {
        "textDecoration": "none",
      }
    }
  >
    Back to procedures
  </a>
  <Loader
    backgroundStyle={
      Object {
        "backgroundColor": "white",
      }
    }
    foregroundStyle={
      Object {
        "color": "black",
      }
    }
    hideContentOnLoad={false}
    message={
      <div
        style={
          Object {
            "borderRadius": 0,
            "justifyContent": "center",
            "left": 0,
            "margin": "auto",
            "position": "absolute",
            "right": 0,
            "top": "-20px",
            "zIndex": 100,
          }
        }
      >
        <DataSpinner
          message="Starting procedure..."
        />
      </div>
    }
    priority={0}
    show={false}
  >
    <div
      className="govuk-grid-row"
    >
      <div
        className="govuk-grid-column-full"
      >
        <div>
          <span
            className="govuk-caption-l"
          >
            Operational procedure
          </span>
          <h2
            className="govuk-heading-l"
          >
            procedure
          </h2>
          <StartForm
            clearProcessDefinition={[MockFunction]}
            dataChange={[Function]}
            fetchProcessDefinition={
              [MockFunction] {
                "calls": Array [
                  Array [
                    "processKey",
                  ],
                ],
                "results": Array [
                  Object {
                    "isThrow": false,
                    "value": undefined,
                  },
                ],
              }
            }
            form={
              Object {
                "components": Array [
                  Object {
                    "allowMultipleMasks": false,
                    "alwaysEnabled": false,
                    "id": "firstNameId",
                    "input": true,
                    "key": "firstName",
                    "label": "First name",
                    "showCharCount": false,
                    "showWordCount": false,
                    "tableView": true,
                    "type": "textfield",
                    "widget": Object {
                      "type": "",
                    },
                  },
                  Object {
                    "allowMultipleMasks": false,
                    "alwaysEnabled": false,
                    "id": "surnameId",
                    "input": true,
                    "key": "surname",
                    "label": "Surname",
                    "showCharCount": false,
                    "showWordCount": false,
                    "tableView": true,
                    "type": "textfield",
                    "widget": Object {
                      "type": "",
                    },
                  },
                  Object {
                    "disableOnInvalid": true,
                    "id": "submitId",
                    "input": true,
                    "key": "submit",
                    "label": "Submit",
                    "tableView": true,
                    "theme": "primary",
                    "type": "button",
                  },
                ],
                "display": "form",
              }
            }
            formReference={[Function]}
            handleSubmit={[Function]}
            history={
              Object {
                "location": Object {
                  "pathname": "/path",
                },
              }
            }
            kc={
              Object {
                "tokenParsed": Object {
                  "email": "email",
                },
              }
            }
            loadingForm={false}
            match={
              Object {
                "params": Object {
                  "processKey": "processKey",
                },
              }
            }
            onCustomEvent={[Function]}
            processDefinition={
              Immutable.Map {
                "formKey": "formKey",
                "process-definition": Immutable.Map {
                  "name": "procedure",
                },
              }
            }
            startForm={
              Object {
                "components": Array [
                  Object {
                    "allowMultipleMasks": false,
                    "alwaysEnabled": false,
                    "id": "firstNameId",
                    "input": true,
                    "key": "firstName",
                    "label": "First name",
                    "showCharCount": false,
                    "showWordCount": false,
                    "tableView": true,
                    "type": "textfield",
                    "widget": Object {
                      "type": "",
                    },
                  },
                  Object {
                    "allowMultipleMasks": false,
                    "alwaysEnabled": false,
                    "id": "surnameId",
                    "input": true,
                    "key": "surname",
                    "label": "Surname",
                    "showCharCount": false,
                    "showWordCount": false,
                    "tableView": true,
                    "type": "textfield",
                    "widget": Object {
                      "type": "",
                    },
                  },
                  Object {
                    "disableOnInvalid": true,
                    "id": "submitId",
                    "input": true,
                    "key": "submit",
                    "label": "Submit",
                    "tableView": true,
                    "theme": "primary",
                    "type": "button",
                  },
                ],
                "display": "form",
              }
            }
            submissionStatus={null}
          />
        </div>
      </div>
    </div>
  </Loader>
</div>
`;
